### PR TITLE
implemented sorting on target reports results view for groups

### PR DESCRIPTION
**Pull request by dsnyder-fairway**

### CODE REVIEW

File: webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts [hunk 1]
```diff
@@ -199,6 +215,42 @@ export class TargetReportComponent implements OnInit, ExportResults {
...
+        : event.data.sort(claimOrdering.on<AggregateTargetScoreRow>(row => row.claim).reverse().compare);
+    } else if (event.field === 'target') {
+      ascending ? event.data.sort(ordering(byString).on<AggregateTargetScoreRow>(row => row.target).compare)
+        : event.data.sort(ordering(byString).on<AggregateTargetScoreRow>(row => row.target).reverse().compare);
+    } else if (event.field === 'subgroup') {
+      // to keep Overall at the top we shift -> sort -> unshift
+      const overall = event.data.shift();
+      ascending ? event.data.sort(SubgroupOrdering.on<AggregateTargetScoreRow>(row => row.subgroup).compare)
+        : event.data.sort(SubgroupOrdering.on<AggregateTargetScoreRow>(row => row.subgroup).reverse().compare);
+      event.data.unshift(overall);
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: why do we need to shift/unshift?

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: keeps Overall at the top with all the other subgroups sorted below

File: webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts [hunk 2]
```diff
@@ -199,6 +215,42 @@ export class TargetReportComponent implements OnInit, ExportResults {
...
   }
 
+  /**
+   * Sort the data
+   *
+   * @param event {{order: number, field: string}} An optional sort event
+   */
+  public sort(event?: SortEvent): void {
+    const ascending = event.order > 0;
+    if (event.field === 'claimOrder') {
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: nit: would add this at the top to remove all the redundant dereferencing
```
const { field, data } = event;
```

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: also a good candidate for a switch statement/ map based lookup like we use the aggregate table.

File: webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts (hunk 1 above)
Comment (wtritch) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: Oh wow... this is really hard to wrap my head around... Is there a better way to break all this up?

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: changes are coming

Comment (esotrope) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: sounds ominous 8 l

File: webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts [hunk 3]
```diff
@@ -199,6 +215,43 @@ export class TargetReportComponent implements OnInit, ExportResults {
...
+   * @param event {{order: number, field: string}} An optional sort event
+   */
+  public sort(event?: SortEvent): void {
+    const { field, data } = event;
+    const ascending = event.order > 0;
+    if (field === 'claimOrder') {
+      // when there isn't a subject specific claim ordering, then default ot a simple alpha sort
+      const claimOrdering: Ordering<string> = (SubjectClaimOrderings.get(this.assessment.subject) || ordering(byString));
+      ascending ? data.sort(claimOrdering.on<AggregateTargetScoreRow>(row => row.claim).compare)
+        : data.sort(claimOrdering.on<AggregateTargetScoreRow>(row => row.claim).reverse().compare);
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: You could simplify all this with something like:
```
let columnOrdering: Ordering<AggregateTargetScoreRow>;
if (field === 'claimOrder') {
  columnOrdering = (SubjectClaimOrderings.get(this.assessment.subject) || ordering(byString))
    .on(row => row.claim);
} else if (blah) {
  ...
}
if (event.order < 0) {
  columnOrdering = columnOrdering.reverse();
}
data.sort(columnOrdering);
```

Comment (dsnyder-fairway) [reply webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: yeah I more or less did this in my latest changes

File: webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts [hunk 4]
```diff
@@ -199,6 +215,48 @@ export class TargetReportComponent implements OnInit, ExportResults {
...
+   * @param event {{order: number, field: string}} An optional sort event
+   */
+  public sort(event?: SortEvent): void {
+    const { field, data } = event;
+    const ascending = event.order > 0;
+    const order = this.getOrdering(field);
+
+    if (field === 'subgroup') {
+      // to keep Overall at the top we shift -> sort -> unshift
+      const overall = data.shift();
```
Comment (wtritch) [inline webapp/src/main/webapp/src/app/assessments/results/view/target-report/target-report.component.ts]: You can do this using a combined Ordering that sorts overall via ranking then falls through to sorting alphabetically.  You can't assume that "overall" will always be the first value when coming into this method (e.g. if they sorted by a different field just prior, overall may be at any position.)

**Comment by dsnyder-fairway** [comment]
@esotrope I think you are right. I was confused about how subgroups are ordered. I'll remove the overall shift / unshift

**Pull request merged**